profiler/compression.go

```diff
@@ -145,3 +191,42 @@ func (r *passthroughCompressor) Reset(w io.Writer) {
 func (r *passthroughCompressor) Close() error {
 	return nil
 }
+
+type zstdRecompressor struct {
+	// err synchronizes finishing writes after closing pw and reports any
+	// error during recompression
+	err     chan error
+	pw      io.WriteCloser
+	zstdOut *zstd.Encoder
+	level   zstd.EncoderLevel
+}
+
+func (r *zstdRecompressor) Reset(w io.Writer) {
+	pr, pw := io.Pipe()
+	// NB: we're assuming the level is valid
+	zstdOut, _ := zstd.NewWriter(w, zstd.WithEncoderLevel(r.level))
+	if r.err == nil {
+		r.err = make(chan error)
+	}
```

Comment:
NIT: I think I slightly prefer having a constructor over lazy initializing things here. It would also be nice to reuse the zstd writer rather than re-creating it on every reset.

Feel free to merge https://github.com/DataDog/dd-trace-go/pull/3572 if that makes sense.

Reply:
Merged, thanks! Your change also lets us check the error from `zstd.NewWriter` 👍